### DISCUSSION TITLE
Configure custom logger on Puma::CLI

### DISCUSSION THREAD

**Discussion by svoop** [puma/puma#3124] [Q&A]
Thanks to @vzajkov with #2770, it's now possible to set a custom logger instance in `puma.rb`. 🎉 

Got a question concerning these custom loggers:

I'm embedding Puma as the backend server in a little framework which features its own CLI for common tasks – one of which is starting and stopping the backend. It pretty much boils down to this:

```
def start(daemonize:)
  require "puma/cli"
  Daemons.daemonize if daemonize
  Puma::CLI.new(arguments).run
end

private

def arguments
  [
    '--dir', MyApp.root.to_s,
    '--bind', "tcp://#{MyApp.bind}",
    (...)
    '--no-config'
  ]
end
```

Obviously, the custom logger instance cannot be passed as an argument, so I took a dive into the source and found out the CLI constructor accepts a `LogWriter`. It then looks like so:

```ruby
def start(daemonize:)
  require "puma/cli"
  Daemons.daemonize if daemonize
  Puma::CLI.new(arguments, log_writer).run
end

private

def log_writer
  Puma::LogWriter.new(STDOUT, STDERR).tap do |log_writer|
    log_writer.custom_logger = MyApp.logger
  end
end
```

Unfortunately, this does not work, logging still happens on STDOUT.

Is there a proper way to do this?

**Comment by nateberkopec** [comment]
I don't think there is a way to do this. What interface would you like/did you expect?

**Reply by svoop** [reply]
Given my use case described above (a littel WIP [chatops framework called Rodbot](https://github.com/svoop/rodbot)), I'm not really sure whether wrapping around `Puma::CLI` is the best way to go, I've essentially peeked how other frameworks achieve this.

If it is, then an additional kwarg on the `Puma::CLI` initializer would be great:

```ruby
Puma::CLI.new(arguments, custom_logger: MyApp.loger).run
```

Such kwargs could cover what cannot be done with `arguments` where CLI arguments are passed as string and therefore are limited to primitives.

However, I'd be happy with the approach described above to work as well. Right now, I don't understand the inner workings well enough to see why passing a custom `Puma::LogWriter` instance this way doesn't work.

**Reply by nateberkopec** [reply]
I think most other places (i.e. Capybara) use the Puma::Server.new interface rather than Puma::CLI, can you do that instead?

**Reply by svoop** [reply]
@nateberkopec Good idea! I've switched to `Puma::Server`, the actual serving (and access log) wasn't difficult:

```
Puma::Server.new(app, nil, options).tap do |server|
  server.add_tcp_listener(*bind)
  server.app = Rack::CommonLogger.new(app, logger)
end.run.join
```

(`logger` is an instance of `Logger` and `app` the result of `Rack::Builder` parsing config.ru)

However, I just don't find a way to have Puma log errors to that `logger` as well.

I can set a custom `LogWriter` as part of the options hash. But `LogWriter` expects an IO, so it won't work with a `Logger` instance as is. A simple `LoggerIO` wrapper might do...

```
def options
  { log_writer: log_writer }
end

def log_writer
  logger_io = LoggerIO.new(logger, Logger::ERROR)
  Puma::LogWriter.new(logger_io, logger_io)
end

class LoggerIO
  def initialize(logger, level)
    @logger, @level = logger, level
  end

  def write(message)
    @logger.log(@level, message.strip)
  end

  def method_missing(*)
    # swallow sync, flush etc
  end
end
```

Looks smart, but doesn't work. I'm certainly missing something here, custom internal logging shouldn't be so hard. Any idea where I'm taking a wrong turn here?

**Reply by svoop** [reply]
@nateberkopec Well, I did find a solution of sorts:

```
def options
  {
    lowlevel_error_handler: method(:lowlevel_error_handler),
    log_writer: Puma::LogWriter.null,
    min_threads: Rodbot.config(:app, :threads).min,
    max_threads: Rodbot.config(:app, :threads).max
  }
end

def lowlevel_error_handler(error)
  logger.error("#{error.message}: #{error.backtrace.first}")
  [500, {}, ['Oops!']]
end
```

I'm silencing the normal logging with `LogWriter.null` and then plug in a custom error handler. Looks like a hack, but since I wanted a custom 500 error page which is not dumping the backtrace for any environment, there's not much extra boilerplate.

Overall, [it looks like this](https://github.com/svoop/rodbot/blob/main/lib/rodbot/services/app.rb). No monkey patching Puma, therefore okay with me. But if you have a better idea, I'm all ears!
